index.rst

```diff
@@ -1,42 +1,198 @@
+*punning*.  Haskell makes heavy use of punning in its built-in syntax and common
+types.
+
+However, there are various contexts in which an occurrence of a name may refer
+either to the type namespace or the data namespace, and it is not always clear
+which is meant. In particular, the following may mention both type-level and
+term-level entities:
+
+- Import and export lists
+
+- Fixity declarations
+
+- ``WARNING``, ``DEPRECATED`` and ``ANN`` pragmas
+
+- Types, when using the ``DataKinds`` extension to reference a data constructor
+  at the type level
+
+- Template Haskell name quotes
+
+In certain situations, which appear to be geining prevalence in modern Haskell, precison is needed
+in specifying exactly which names are exported or imported from a module. This extension aims to
+provide precise control in the presence of puns where the same name is present in the type and
+data namespace.
+
+ExplicitNamespaces Overview
+---------------------------
+
+When the ``ExplicitNamespaces`` extensiopn is enabled, ``type`` and ``data`` keywords may be prefixed
+to a qualified constructor names in import and export lists to specify whether a type constructor
+or data constructor is being exported or brought into scope.  It is an error to use such a namespace
+specifier if the identifier is not in scope in the given namespace.
+
+More precisely, the standard grammar of Haskell import/export items is
+essentially the following (after some minor simplifications): ::
+
+      export -> qcname_ext ['(' qcname_ext_w_wildcard_1, ..., qcname_ext_w_wildcard_n ')']
+             |  'module' modid
+
+      qcname_ext_w_wildcard -> qcname_ext
+                            |  '..'
+
+      qcname_ext -> qvar
+                 |  qtycon
+
+This extension extends ``qcname_ext`` as follows: ::
+
+      qcname_ext -> qvar
+                 |  qtycon
+                 | 'type' oqtycon  -- with ExplicitNamespaces
+                 | 'data' qvarcon  -- with ExplicitNamespaces
+
+Notice that:
+
+- ``module`` is valid only at the top level of the export,
+  whereas ``type`` and ``data`` are valid either at the top or nested inside a
+  type constructor or typeclass name.
+
+- ``data`` may be followed by a data constructor name or a variable name (with
+  the latter including record selectors, in particular).
+
+
+Namespace-specified imports
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+When the ``ExplicitNamespaces`` extension is enabled, the syntax of import
+declarations is extended to include a namespace specifier immediately after the
+module identifier.
+
+More concretely, in the grammar accepted by GHC, ::
+
+      importdecl -> 'import' [src] ['safe'] ['qualified'] [package] modid ['qualified'] ['as' modid] [impspec]
+
+is changed to ::
+
+      importdecl -> 'import' [src] ['safe'] ['qualified'] [package] modid [namespace] ['qualified'] ['as' modid] [impspec]
+
+      namespace  -> 'data'
+                 |  'type'
```

Comment:
Again, explicit grammars feel a bit heavy for the user's guide.

Reply:
Likewise, I think we should be documenting how we are modifying the grammar.